java/core/src/main/java/co/worklytics/psoxy/gateway/output/ApiDataOutputUtils.java

```diff
@@ -0,0 +1,267 @@
+    Base64.Encoder base64encoder;
+
+
+    public String buildRawOutputKey(HttpRequest requestToSourceApi) {
+        Map<String, String> metadata = buildRawMetadata(requestToSourceApi);
+        return buildRawOutputKey(metadata);
+    }
+
+    public String buildRawOutputKey(ProcessedContent content) {
+        return buildRawOutputKey(content.getMetadata());
+    }
+
+    private String buildRawOutputKey(Map<String, String> metadata) {
+
+        String host = metadata.get(OutputObjectMetadata.API_HOST.name());
+        String path = metadata.get(OutputObjectMetadata.PATH.name());
+
+        String pathSegment = Optional.ofNullable(path).map(s -> "/" + s.replace("/", "_")).orElse("");
+
+        return host + pathSegment + "/" + uuidProvider.get().toString();
+    }
+
+    public String buildSanitizedOutputKey(HttpEventRequest requestToProxy) {
+        //TODO: host here is going to be the proxy host, not the source API host.
+        //TODO: path may include some additional segments, unless properly stripped by adapters before call into ApiDataHandler ...
+        Map<String, String> metadata = buildMetadata(requestToProxy);
+        return buildSanitizedOutputKey(metadata);
+    }
+
+    public String buildSanitizedOutputKey(ProcessedContent content) {
+        return buildSanitizedOutputKey(content.getMetadata());
+    }
+
+    private String buildSanitizedOutputKey(Map<String, String> metadata) {
+
+        String host = metadata.get(OutputObjectMetadata.API_HOST.name());
+        String path = metadata.get(OutputObjectMetadata.PATH.name());
+
+        String sanitizedPath = Optional.ofNullable(path)
+            .map(this::normalizePath)
+            .map(s -> s.replace("/", "_") + "/")
+            .orElse("");
+
+        return host + "/" + sanitizedPath + uuidProvider.get().toString();
+    }
+
+
+    /**
+     * wrap api response produced by request as {@link ProcessedContent}
+     *
+     * in effect, it's the request that's been processed.
+     *
+     *
+     * @param sourceApiRequest request to the source API
+     * @param sourceApiResponse response from the source API
+     * @return ProcessedContent containing the response content, metadata, etc.
+     * @throws IOException
+     */
+    public ProcessedContent responseAsRawProcessedContent(HttpRequest sourceApiRequest,
+                                                          HttpResponse sourceApiResponse) throws IOException {
+        ProcessedContent.ProcessedContentBuilder builder = ProcessedContent.builder()
+            .contentType(sourceApiResponse.getContentType())
+            .contentCharset(sourceApiResponse.getContentCharset());
+
+        try (InputStream stream = sourceApiResponse.getContent()) {
+            builder.content(stream.readAllBytes());
+        }
+
+        Map<String, String> metadata = this.buildRawMetadata(sourceApiRequest);
+
+
+        //not sure this will work; are we certain to be able to consume HttpContent after request has been sent?
```

Comment:
I think not, we'll need to *store* the content in some kind of buffered stream to be reused, otherwise the 2nd time content is read it will fail